tensorboard/summary/writer/event_file_writer.py

```diff
@@ -218,6 +240,22 @@ def stop(self):
         self.join()
 
     def run(self):
+        try:
+            self._run()
+        except Exception as ex:
+            self.exception = ex
+            try:
+                # In case there's a thread blocked on putting an item into the
+                # queue, pop one item and immediatelly mark it as done. Note
+                # that since the worker status check is done under a lock,
+                # there's no need to pop more than one item.
+                self._queue.get(False)
```

Comment:
While this does seem to handle the case of deadlock for the `write` case, I don't believe this will suffice for `flush` (since `Queue.join()` doesn't have a timeout) or `stop`/`close` since they need the entire entire queue cleared to unblock the main thread.

Reply:
Good point, for `join` it might not always work, replaced popping a single item with emptying the whole queue. Thanks for noticing and pointing out!

I believe it should be enough, since flush and write happen under the same lock. If at the moment `self.exception` is set the flush critical section is being executed, then no write will interfere with emptying the queue. If at the moment of setting `self.exception` flush critical section is not yet being executed, then as soon as the lock is acquired, it will fail after the check (not that I moved the check under the lock)

`close` doesn't flush the queue, it flushed the underlying reader

Did you have something else in mind?

Reply:
Thanks, this looks like it should handle `write` and `flush`!

For `close`, I had originally thought deadlock could happen since it calls `self._worker.stop()` which places the shutdown signal in the queue and calls `join` on the worker thread. However, that call is (1) nonblocking on the main thread, (2) made under the lock, and (3) sets `self._closed` to True, so any subsequent calls to `write` or `flush` should result in an `IOError` before interacting with the queue again. So that should be safe as well.